src/lib/utils/meta_tables/meta_cached_operators_table.cpp

```diff
@@ -0,0 +1,61 @@
+#include "meta_cached_operators_table.hpp"
+
+#include "hyrise.hpp"
+
+namespace opossum {
+
+MetaCachedOperatorsTable::MetaCachedOperatorsTable()
+    : AbstractMetaTable(TableColumnDefinitions{{"operator", DataType::String, false},
+                                               {"query_hash", DataType::String, false},
+                                               {"description", DataType::String, false},
+                                               {"walltime_ns", DataType::Long, false},
+                                               {"output_chunks", DataType::Long, false},
+                                               {"output_rows", DataType::Long, false}}) {}
+
+const std::string& MetaCachedOperatorsTable::name() const {
+  static const auto name = std::string{"cached_operators"};
+  return name;
+}
+
+std::shared_ptr<Table> MetaCachedOperatorsTable::_on_generate() const {
+  auto output_table = std::make_shared<Table>(_column_definitions, TableType::Data, std::nullopt, UseMvcc::Yes);
+  if (!Hyrise::get().default_pqp_cache) return output_table;
+
+  for (auto iter = Hyrise::get().default_pqp_cache->unsafe_begin();
+       iter != Hyrise::get().default_pqp_cache->unsafe_end(); ++iter) {
```

Comment:
@Bouncner If we want to use that for the BP, we need safe iterators here, don't we? We could introduce a lock or use the recently discussed tab::concurrent_unordered_map? Any other thoughts or comments?

Reply:
Uff ... good question. No idea to be honest.

Reply:
I think @mrks is correct: if we insert into the cache, which should happen all the time during the BP scenario, this could potentially crash when we iterate at the same time which will happen simultaneously in the BP scenario.

Therefore, I don't see a way around implementing a safe way to iterate the cache. As we have just validated that the tbb::concurrent_unordered_map is often faster as a locking approach, we could go for it.

Reply:
As far as I have understood it, the different Cache implementations use different data structures (e.g. a Fibonacci heap ad priority queue in GDFSCache). Are there TBB structures for all those implementations (or at least for the default cache) that support the functionality currently used? If not, we can only lock the caches, right?

Reply:
After we discussed providing a snapshot copy of the GDFS cache data earlier, I again had a look at the data structures. My first attempt was to copy the GDFS cache map, which contains handles to the queued entries. As this was part of GDFS cache's internals, I needed to add locks wherever its internal functions touch that data ([see here](https://github.com/hyrise/hyrise/compare/dey4ss/lqp_cache_meta#diff-e31d51d3d7bbe1d603e6339356d7c3f9)).
Anyhow, this leads to adding locks in every cache operation function and re-implements the [locks of `cache.hpp`](https://github.com/hyrise/hyrise/blob/master/src/lib/cache/cache.hpp) (but can be done on a lower and hence finer level).
As this "double locking" does not seem to be right, I came up with two ideas:
1. Add a snapshot method to `cache.hpp` and `AbstractCacheImpl`. The method is locked by cache.hpp and then forwarded to the implementation, which returns a map of entries with a defined format.
2. Move all the locking to the implementations. Besides the little more effort, I see some advantages in doing that:
2.1 We could gain performance: Instead of locking the whole operation, we can lock only the specific data structure accesses.
2.2 We would let the implementation decide how to achieve thread-safeness. Especially when achievable with data structures (e.g. tbb concurrent map), this is clearly an advantage.
2.3 We could get rid of the Cache and CacheImpl hierarchy. The only difference is locking, so we could trim the architecture a little.

As we currently only use the GDFS cache, we then could use a concurrent map as CacheMap and only lock the queue. As I am not completely sure how concurrent map iterations or find() works, it might be necessary to have a lock for each entry, too. Anyhow, we should gain performance, as we do not need to lock the whole operation. 
For any other cache, locking the whole operations is easy and not worse than the current approach (as we do exactly that now). Maybe we can even get around this by using concurrent maps.

For snapshots, we can also make use of the concurrent unordered map. If the calculations inside the loop do not take long, we could use its iterator, which should be thread-safe. Otherwise, we can keep copying the map.

As this is a little work I already suggested to @Bensk1 and @Bouncner, I would be glad to get some input on that, e.g. from @mrks.

Reply:
I think the original idea was to decouple the cache implementation from the locking. However, your arguments make sense and I don't have much to add, would "Option 2 sounds better" be an acceptable answer?

Reply:
Not only acceptable, but the answer I wanted to hear ;)